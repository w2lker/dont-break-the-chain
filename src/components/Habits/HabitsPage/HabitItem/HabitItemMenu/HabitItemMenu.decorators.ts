// tslint:disable:no-console

import withStyles from '@material-ui/core/styles/withStyles';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';

import HabitItemMenu, { IHabitItemMenuProps } from './HabitItemMenu';
import habitItemMenuStyles from './HabitItemMenu.styles';

const mapStateToProps = (state: any): IHabitItemMenuProps => {
  return {} as IHabitItemMenuProps;
};

const mapDispatchToProps = (dispatch: ThunkDispatch<any, any, any>): IHabitItemMenuProps => {
  return {
    onCheck: () => console.log('-----', 'TODO: Add check function'),
    onUncheck: () => console.log('-----', 'TODO: Add uncheck function'),
    onPause: () => console.log('-----', 'TODO: Add pause function'),
    onDelete: () => console.log('-----', 'TODO: Add delete function'),
  } as IHabitItemMenuProps;
};

const connectedComponent = connect(mapStateToProps, mapDispatchToProps)(HabitItemMenu);

const styledComponent = withStyles(habitItemMenuStyles)(connectedComponent);

// tslint:enable:no-console
export default styledComponent;
