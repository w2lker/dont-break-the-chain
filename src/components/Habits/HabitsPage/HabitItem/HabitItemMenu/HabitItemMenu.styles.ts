const habitItemMenuStyles = {
  wrapper: {
    marginLeft: 'auto',
  },
};

export default habitItemMenuStyles;
