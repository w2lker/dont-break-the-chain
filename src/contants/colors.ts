import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
