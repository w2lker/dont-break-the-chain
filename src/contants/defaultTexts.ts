const defaultTexts = {
  navigationMenu: {
    chains: 'Chains',
    habits: 'Habits',
    profile: 'Profile',
  },
  habits: {
    header: {
      title: 'My Habits',
    },
  },
  profile: {
    header: {
      title: 'Profile',
    },
    stats: {
      total: {
        description: 'points received',
      },
      days: {
        subtitle: 'days',
        description: 'longest chain',
      },
    },
  },
};

export default defaultTexts;
