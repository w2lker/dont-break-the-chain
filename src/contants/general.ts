import { RequestStatuses } from '../models/general';

export const requestStatuses = {
  active: 'active',
  success: 'success',
  error: 'error',
};
