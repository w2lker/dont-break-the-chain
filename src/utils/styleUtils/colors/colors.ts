import {bonWarning} from '../helpers/warning/warning';

export interface IRgbColor {
  red: number;
  green: number;
  blue: number;
}

const black: IRgbColor = {
  red: 0,
  green: 0,
  blue: 0,
};

function regularHexToRgb(hex: string): IRgbColor {
  const [red, green, blue] = [0, 2, 4]
    .map(offset => hex.substring(offset, offset + 2))
    .map(hexCh => parseInt(hexCh, 16));
  return { red, green, blue };
}

export function hexToRgbMapped(hex?: string): IRgbColor {
  const warnType = 'HEX to RGB converter';
  if (!hex) {
    return black;
  }
  if (hex.length === 1) {
    return regularHexToRgb(hex.repeat(6));
  }

  if (hex.length === 2) {
    return regularHexToRgb(hex.repeat(3));
  }

  if (hex.length === 3) {
    const [r, g, b] = hex.split('') as string[];
    return regularHexToRgb(`${r}${r}${g}${g}${b}${b}`);
  }

  if (hex.length === 4 || hex.length === 5) {
    const hex3 = hex.slice(0, 3);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex3} value`, warnType);
    return hexToRgbMapped(hex3);
  }

  if (hex.length > 6) {
    const hex6 = hex.slice(0, 6);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex6} value`, warnType);
    return regularHexToRgb(hex6);
  }

  return regularHexToRgb(hex);
}
