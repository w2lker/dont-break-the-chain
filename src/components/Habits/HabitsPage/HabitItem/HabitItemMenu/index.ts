import HabitItemMenu from './HabitItemMenu.decorators';

export default HabitItemMenu;
