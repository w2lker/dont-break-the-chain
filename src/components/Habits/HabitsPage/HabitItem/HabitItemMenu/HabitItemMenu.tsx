// tslint:disable:no-console
import React, { useRef, useState } from 'react';

import { IconButton, Menu } from '@material-ui/core';
import { List } from 'immutable';
import { WithStyles } from 'react-jss';

import MaterialDesignIcon from '../../../../basic/Material-Icon';
import HabitItemMenuContent from '../HabitItemMenuContent';
import habitItemMenuStyles from './HabitItemMenu.styles';

import { routing } from '../../../../../contants/routing';
import { habitStatus, IHabit } from '../../../../../models/habit';

export interface IHabitItemMenuProps extends WithStyles<typeof habitItemMenuStyles> {
  content?: IHabit;
  onCheck: () => void;
  onUncheck: () => void;
  onPause: () => void;
  onDelete: () => void;
}

const HabitItemMenu: React.FC<IHabitItemMenuProps> = (props) => {
  const {
    content,
    classes,
    onCheck,
    onUncheck,
    onPause,
    onDelete,
  } = props;

  const [menuOpened, setMenuOpened] = useState(false);
  const menuHolderRef = useRef<HTMLButtonElement>(null);

  if (!content) {
    return  null;
  }

  const habitsRouting = routing.habits;
  const { id, dates } = content as IHabit;
  // @ts-ignore
  const latestDate = dates && List.isList(dates) && dates.get(0);
  const isTodayCompleted = latestDate && latestDate.status === habitStatus.end;
  const linkToUpdate = id ? habitsRouting.root + habitsRouting.update.slice(-2) + id : '';

  return (
    <div className={classes.wrapper}>
      <IconButton
        ref={menuHolderRef}
        onClick={() => setMenuOpened(!menuOpened)}
      >
        <MaterialDesignIcon name="dots-vertical" />
      </IconButton>
      <Menu
        id={`habit-menu-${id}`}
        anchorEl={menuHolderRef.current}
        keepMounted
        open={menuOpened}
        onClose={ () => setMenuOpened(false) }
      >
        {isTodayCompleted && (
          <HabitItemMenuContent
            icon="minus"
            title="Unmark today"
            action={onUncheck}
          />
        )}

        {!isTodayCompleted && (
          <HabitItemMenuContent
            icon="check"
            title="Complete today"
            action={onCheck}
          />
        )}

        <HabitItemMenuContent
          icon="pencil"
          title="Edit"
          url={linkToUpdate}
        />
        <HabitItemMenuContent
          icon="pause"
          title="Pause"
          action={onPause}
        />

        <HabitItemMenuContent
          icon="delete"
          title="Delete"
          action={onDelete}
        />
      </Menu>
    </div>
  );
};

// tslint:enable:no-console
export default HabitItemMenu;
