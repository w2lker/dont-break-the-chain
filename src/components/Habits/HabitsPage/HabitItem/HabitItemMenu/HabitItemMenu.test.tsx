import cases from 'jest-in-case';
import React from 'react';

import { mount, shallow } from 'enzyme';
import { List } from 'immutable';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { colors } from '../../../../../contants/colors';
import HabitItemMenu, { IHabitItemMenuProps } from './HabitItemMenu';
import HabitItemMenuDecorated from './HabitItemMenu.decorators';
import habitItemMenuStyles from './HabitItemMenu.styles';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const setupTests = () => {
  const classes = {
    wrapper: 'habitItemMenu-wrapper',
  };
  const classesKeys = Object.keys(classes);
  const emptyProps: IHabitItemMenuProps = {
    classes,
    onCheck: jest.fn(),
    onUncheck: jest.fn(),
    onPause: jest.fn(),
    onDelete: jest.fn(),
  };

  const sampleProps: IHabitItemMenuProps = {
    ...emptyProps,
    content: {
      id: 1,
      name: 'Sample habit',
      color: colors.blue,
      icon: 'sample-icon',
      currentChain: 1,
      longestChain: 2,
      dates: List(),
    },
  };

  const propsCleaner = {
    classes: null,
    onCheck: null,
    onUncheck: null,
    onPause: null,
    onDelete: null,
  };

  const wrappedComponent = (props: IHabitItemMenuProps) => (
    <MemoryRouter>
      <HabitItemMenu {...props} />
    </MemoryRouter>
  );

  const store = mockStore({});

  const wrappedComponentDecorated = (props: IHabitItemMenuProps) => (
    <MemoryRouter>
      <Provider store={store}>
        <HabitItemMenuDecorated {...props} />
      </Provider>
    </MemoryRouter>
  );

  return { classes, classesKeys, emptyProps, sampleProps, propsCleaner, wrappedComponentDecorated };
};

describe('HabitItemMenu component', () => {

  it('match snapshot', () => {
    const { sampleProps } = setupTests();
    const component = shallow(<HabitItemMenu {...sampleProps} />);
    expect(component.debug()).toMatchSnapshot();
  });

  it('skips render with empty props', () => {
    const { emptyProps } = setupTests();
    const component = shallow(<HabitItemMenu {...emptyProps} />);
    expect(component.type()).toEqual(null);
  });

  const menuCasesTitle = 'Renders checked menu';
  const menuCasesTester = (opts: typeof menuCasesData[0]) => {
    const { sampleProps } = setupTests();
    const component = shallow(<HabitItemMenu {...sampleProps} />);
    // @ts-ignore
    const menu = component.find(`#habit-menu-${sampleProps.content.id}`).children();
    const menuItemProps = menu.get(opts.index);
    expect(menuItemProps.props.icon).toBe(opts.icon);
  };

  const menuCasesData = [
    { name: 'Status menu item is rendered', index: 0, icon: 'check' },
    { name: 'Edit menu item is rendered', index: 1, icon: 'pencil' },
    { name: 'Pause menu item is rendered', index: 2, icon: 'pause' },
    { name: 'Delete menu item is rendered', index: 3, icon: 'delete' },
  ];

  cases(menuCasesTitle, menuCasesTester, menuCasesData);

});

describe('HabitItemMenu styles', () => {
  it('contains principle classes', () => {
    const { classesKeys } = setupTests();
    expect(habitItemMenuStyles).toBeDefined();
    classesKeys.forEach((keyValue) => {
      // @ts-ignore
      expect(habitItemMenuStyles[keyValue]).toBeDefined();
    });
  });
});

describe('HabitItemMenu decorators', () => {
  test.todo('provides classes to the component');
});
