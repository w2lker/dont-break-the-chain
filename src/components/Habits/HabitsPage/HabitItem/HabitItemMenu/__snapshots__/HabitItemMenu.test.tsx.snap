// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitItemMenu component match snapshot 1`] = `
"<div className=\\"habitItemMenu-wrapper\\">
  <WithStyles(ForwardRef(IconButton)) onClick={[Function: onClick]}>
    <MaterialDesignIcon name=\\"dots-vertical\\" />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(Menu)) id=\\"habit-menu-1\\" anchorEl={{...}} keepMounted={true} open={false} onClose={[Function: onClose]}>
    <WithStyles(HabitItemMenuContent) icon=\\"check\\" title=\\"Complete today\\" action={[Function: mockConstructor]} />
    <WithStyles(HabitItemMenuContent) icon=\\"pencil\\" title=\\"Edit\\" url=\\"/habitid1\\" />
    <WithStyles(HabitItemMenuContent) icon=\\"pause\\" title=\\"Pause\\" action={[Function: mockConstructor]} />
    <WithStyles(HabitItemMenuContent) icon=\\"delete\\" title=\\"Delete\\" action={[Function: mockConstructor]} />
  </WithStyles(ForwardRef(Menu))>
</div>"
`;
